### PR TITLE
Emoji 4.0 support & only preprocessing emoji shortcodes

### PR DESCRIPTION
**Pull request by ticky**
Reverts buildkite/frontend#272, restoring buildkite/frontend#270.

### CODE REVIEW

File: app/css/emoji.css [hunk 1]
```diff
@@ -1,5 +1,13 @@
...
-  width: 1.25em;
-  height: 1.25em;
+  width: 1.23em;
+  height: 1.23em;
+  margin-left:.05em;
+  margin-right:.05em;
   vertical-align: -.2em;
 }
+
+.tiny-kitemoji .emoji {
```
Comment (keithpitt) [inline app/css/emoji.css]: Kitemoji! Possible the best name ever?

File: app/components/shared/Emojify.js [hunk 2]
```diff
@@ -3,25 +3,35 @@ import PropTypes from 'prop-types';
 
 import Emoji from '../../lib/Emoji';
 
-class Emojify extends React.PureComponent {
+export default class Emojify extends React.PureComponent {
```
Comment (sj26) [inline app/components/shared/Emojify.js]: Making this pure seems like it could speed things up a bit :clap:

File: app/lib/Emoji.js [hunk 3]
```diff
@@ -1,78 +1,49 @@
...
-  }
-
-  _image(catalogue, emoji) {
-    // Emoji catalogue hosts have a normalized host that always end with a "/"
-    const emojiUrl = `${catalogue.host}${emoji.image}`;
-    const emojiCanonicalRepresentation = emoji.unicode || `:${emoji.name}:`;
+      return match;
 
-    return `<img class="emoji" title="${emoji.name}" alt="${emojiCanonicalRepresentation}" src="${emojiUrl}" draggable="false" />`;
+    });
```
Comment (sj26) [inline app/lib/Emoji.js]: Nice. :100:

One of the best performance tweaks I made to the backend emoji processing code was inverting the loops, changing to a single scan over the text which then looped through the catalogues to find matches. Is that something which might help here? I know they behave slightly differently here but that behaviour could be combined.

Comment (ticky) [inline app/lib/Emoji.js]: It sure could be, yeah!

**Comment by ticky** [comment]
Bonus points: this now cuts the client-side Apple emoji catalogue size down by about ⅓ by removing the unicode version of the index, as well as all the image URLs (as they’re no longer used). We could go further if we’re convinced we’re not going to be landing that custom emoji picker anytime soon. :wink:

**Comment by ticky** [comment]
The bug I was trying to mitigate by adding server-side support to this showed up in testing. After a bit of research, I ended up landing ticky/mojibaka#8, which allowed the tests to be moved to happen immediately in app.js’ context, rather than after the entire page loads, this seems to mitigate it properly!

**Pull request merged**